sklearn/model_selection/tests/test_validation.py

```diff
@@ -777,6 +778,64 @@ def split(self, X, y=None, groups=None):
     assert_raises(ValueError, cross_val_predict, est, X, y, cv=BadCV())
 
 
+def test_cross_val_predict_decision_function_shape():
+    X, y = make_classification(n_classes=2, n_samples=50)
+
+    preds = cross_val_predict(LogisticRegression(), X, y,
+                              method='decision_function')
+    assert_equal(preds.shape, (50,))
+
+    X, y = load_iris(return_X_y=True)
+
+    preds = cross_val_predict(LogisticRegression(), X, y,
+                              method='decision_function')
+    assert_equal(preds.shape, (150, 3))
+
+    # This specifically tests imbalanced splits for binary
+    # classification with decision_function. This is only
+    # applicable to classifiers that can be fit on a single
+    # class.
+    X = X[:100]
+    y = y[:100]
+    preds = cross_val_predict(RidgeClassifier(), X, y,
+                              method='decision_function', cv=KFold(2))
+    assert_equal(preds.shape, (100,))
```

Comment:
Can we somehow check that the signs here are correct, not just the shape?

Reply:
I can't seem to comment on the review just above this (Github bug?), so I'll put my reply in here.

> I'm a little bit confused by this if else sequence. Why do we need the first if in addition to the second if? doesn't that cover all cases?

There is a lot of special casing here so not sure what you mean by "first if" and "second if" but I'll try to explain anyway.

```python
if method == 'decision_function' and n_classes == 2:
    predictions_ = predictions_[:, 1]
```

pertains to the case you mentioned "when some classifiers have 1 class and some have 2 classes." Without this line, the output shape would be `(n_samples, 2)` instead of `(n_samples,)`. This is only applicable to `decision_function`. It forces `predictions_` into the correct shape.

If the first if you pertain to is 
```python
if method == 'decision_function' and len(estimator.classes_) <= 2:
    predictions_[:, estimator.classes_[-1]] = predictions
```
then this is for handling when a particular CV fold has <2 classes, and we need to handle it properly when building `predictions_`. 

> Can we somehow check that the signs here are correct, not just the shape?

Not sure what you mean by signs.

Reply:
Is there a case with ``n_classes == 2`` where ``len(estimator.classes_) > 2``? It looks to me like these two ``if``s are always either both true or neither but I might be missing something.

By sign, I mean that in the two-class case, the meaning of the decision function changes depending on whether the one class is the positive class or the negative class, right? If the decision function in the one-class case is always 0, I guess it doesn't make a difference, but I haven't checked that.

Reply:
> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`

> By sign, I mean that in the two-class case, the meaning of the decision function changes depending on whether the one class is the positive class or the negative class, right? If the decision function in the one-class case is always 0, I guess it doesn't make a difference, but I haven't checked that.

For `RidgeClassifier`, it's actually -1. I'm not sure what that means and it's not a regression so I didn't attempt to fix it here. That said, I'm not really sure what a "correct" implementation would look like.

Reply:
I'll double check. Can you comment on the first part? Is there a case when only one of the two if's is true?

Reply:
Can you provide the code you used for RidgeClassifier?

Reply:
If you check the "Files changed" tab, you'll actually see my comment on the first part. Weird Github bug doesn't show it here I guess.. So i'll copy it here.

> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`

Reply:
Here's the code I used for `RidgeClassifier`. 

```python
from sklearn.datasets import load_iris
X,y = load_iris(True)
from sklearn.linear_model import RidgeClassifier
from sklearn.model_selection import cross_val_predict
cross_val_predict(RidgeClassifier(), X[:20], y[:20], method='decision_function')
```
```
array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1.])
```
this is v0.18